ngraph_bridge/ngraph_data_cache.h

```diff
@@ -0,0 +1,178 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************/
+
+#ifndef NGRAPH_DATA_CACHE_H_
+#define NGRAPH_DATA_CACHE_H_
+#pragma once
+
+#include <deque>
+#include <mutex>
+#include <ostream>
+#include <vector>
+#include "absl/synchronization/mutex.h"
+
+#include "tensorflow/core/framework/tensor_shape.h"
+#include "tensorflow/core/graph/graph.h"
+
+#include "logging/ngraph_log.h"
+#include "ngraph/ngraph.hpp"
+
+#include "ngraph_bridge/ngraph_freshness_tracker.h"
+#include "ngraph_bridge/ngraph_pipelined_tensors.h"
+
+namespace tensorflow {
+
+namespace ngraph_bridge {
+
+// Forward declaration for friend class
+namespace testing {
+class NGraphDataCacheTest_SameKeyMultiThread_Test;
+class NGraphDataCacheTest_RemoveItemTest_Test;
+}
+
+template <typename KeyType, typename ValueType>
+class NgraphDataCache {
+ public:
+  explicit NgraphDataCache(int depth);
+  ~NgraphDataCache();
+
+  // This method performs lookup in the cache for requested key, if not found
+  // it will create item, put it in the cache and returns item and status
+  std::pair<Status, ValueType> LookUpOrCreate(
+      KeyType key,
+      std::function<std::pair<Status, ValueType>()> callback_create_item,
+      std::function<void(ValueType)> callback_destroy_item);
+
+  // Overload for above function, which doesn't take callback for destroy item
+  std::pair<Status, ValueType> LookUpOrCreate(
+      KeyType key,
+      std::function<std::pair<Status, ValueType>()> callback_create_item);
+  void RemoveItem(KeyType key);
+  void RemoveItem(KeyType key,
+                  std::function<void(ValueType)> callback_destroy_item);
+  void RemoveAll(std::function<void(ValueType)> callback_destroy_item);
+
+ private:
+  std::unordered_map<KeyType, ValueType> m_ng_items_map;
+  std::deque<KeyType> m_lru;
+  int m_depth;
+  absl::Mutex m_mutex;
+
+  // Test class
+  friend class tensorflow::ngraph_bridge::testing::
+      NGraphDataCacheTest_SameKeyMultiThread_Test;
+  friend class tensorflow::ngraph_bridge::testing::
+      NGraphDataCacheTest_RemoveItemTest_Test;
+};
+
+template <typename KeyType, typename ValueType>
+NgraphDataCache<KeyType, ValueType>::NgraphDataCache(int depth)
+    : m_depth(depth) {}
+
+template <typename KeyType, typename ValueType>
+NgraphDataCache<KeyType, ValueType>::~NgraphDataCache() {
+  m_ng_items_map.clear();
+}
+
+template <typename KeyType, typename ValueType>
+void NgraphDataCache<KeyType, ValueType>::RemoveItem(
+    KeyType key, std::function<void(ValueType)> callback_destroy_item) {
+  if (m_ng_items_map.find(key) != m_ng_items_map.end()) {
```

Comment:
Do we need to lock the look-up?

Reply:
I think yes.

Reply:
The overload should not need to do the lookup etc.
1. RemoveItem(KeyType key, std::function<void(ValueType)> callback_destroy_item) could do all the work
2. RemoveItem(KeyType key) should just call 1. with a dummy function passed for callback_destroy_item

Reply:
Agree, but There is a possibility that this function(The one which takes only key) can be called independent of the other one,
and there is no harm in checking it again